Userland/Games/Sudoku/SudokuWidget.cpp

```diff
@@ -0,0 +1,161 @@
+
+    GUI::Frame::paint_event(event);
+
+    GUI::Painter painter(*this);
+    painter.add_clip_rect(event.rect());
+
+    painter.fill_rect({ 0, 0, width(), height() }, Color::Black);
+
+    painter.translate(frame_thickness() + widget_offset_x,
+        frame_thickness() + widget_offset_y);
+
+    if (m_board) {
+
+        auto board_dimension = m_board->dimension();
+
+        auto tile_size = Gfx::IntSize { m_cell_size * 0.97f, m_cell_size * 0.97f };
+        for (size_t x = 0; x < board_dimension; x++) {
+            for (size_t y = 0; y < board_dimension; y++) {
+
+                // Create subsquare borders
+                if (x % 3 == 0 && y % 3 == 0) {
+                    auto sub_square_size = Gfx::IntSize { m_cell_size * 2.97f, m_cell_size * 2.97f };
+                    Gfx::IntPoint top_left_corner_of_sub_square = { m_cell_size * x,
+                        m_cell_size * y };
+                    Gfx::IntRect sub_square_rect = { top_left_corner_of_sub_square, sub_square_size };
+                    painter.fill_rect(sub_square_rect, Color::LightGray);
+                }
+
+                // Paint tiles
+                // FIXME: Colors should be changeable. Also text isn't very visible
+                // against the active tile shading
+                Gfx::IntPoint tile_location = { m_cell_size * x, m_cell_size * y };
+                Gfx::IntRect tile_rect = { tile_location, tile_size };
+                Square* square = m_board->get_square(x, y);
+                Color square_color = Color::White;
+                if (square == m_active_square) {
+                    square_color = Color::LightGray;
+                }
+                painter.fill_rect(tile_rect, square_color);
+
+                Color text_color = Color::WarmGray;
+                if (square->get_value()) {
+                    if (square->is_fixed()) {
+                        text_color = Color::Black;
+                    }
+                    painter.draw_text(tile_rect, String::number(square->get_value()),
+                        font(), Gfx::TextAlignment::Center, text_color);
+                }
+            }
+        };
+    };
+}
+
+void SudokuWidget::resize_event(GUI::ResizeEvent&)
+{
+    if (m_board) {
+        auto board_dimension = m_board->dimension();
+        m_cell_size = min(width() / (board_dimension), height() / (board_dimension));
+        pick_font();
+    }
+}
+
+void SudokuWidget::mousedown_event(GUI::MouseEvent& event)
+{
+
+    if (!frame_inner_rect().contains(event.position()))
+        return;
+
+    Square* square = mouse_to_square(event);
+    if (m_active_square == square) {
+        m_active_square = nullptr;
+    } else {
+        m_active_square = square;
+    }
+    update();
+}
+
+void SudokuWidget::keydown_event(GUI::KeyEvent& event)
+{
+    if (event.code_point() >= '1' && event.code_point() <= '9') {
+        u32 digit = event.code_point() - '0';
+        m_active_square->set_value(digit);
```

Comment:
Missing a null check here! Pressing any key when you don't have a selected square crashes.